pkg/proc/threads.go

```diff
@@ -325,6 +340,51 @@ func next(dbp Process, stepInto, inlinedStepOut bool) error {
 	return nil
 }
 
+func setStepIntoBreakpoints(dbp Process, text []AsmInstruction, topframe Stackframe, sameGCond ast.Expr) error {
+	for _, instr := range text {
+		if instr.Loc.File != topframe.Current.File || instr.Loc.Line != topframe.Current.Line || !instr.IsCall() {
+			continue
+		}
+
+		if instr.DestLoc != nil && instr.DestLoc.Fn != nil {
+			if err := setStepIntoBreakpoint(dbp, []AsmInstruction{instr}, sameGCond); err != nil {
+				return err
+			}
+		} else {
+			// Non-absolute call instruction, set a StepBreakpoint here
+			if _, err := allowDuplicateBreakpoint(dbp.SetBreakpoint(instr.Loc.PC, StepBreakpoint, sameGCond)); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+var errStepShouldStepInstruction = errors.New("step should step by one instruction")
+
+func setStepIntoBreakpointsReverse(dbp Process, text []AsmInstruction, topframe Stackframe, sameGCond ast.Expr) error {
```

Comment:
It would actually have to be all the RET instructions of every called function. But we don't always know which functions get called for dynamic CALLs (function pointers, interface methods, etc...).

So what this does is it sets a `Kind==StepBreakpoint` breakpoint after every call, then reverse continues and if one of them gets hit, it executes a StepInstruction which takes us on the relevant RET instruction.

This way we also step into any deferred call there might be (just like the forward StepInto steps into deferred calls).

Reply:
Ack, makes sense.